packages/block-library/src/heading/autogenerate-anchors.js

```diff
@@ -0,0 +1,219 @@
+ */
+function getTextWithoutMarkup( text ) {
+	dummyElement.innerHTML = text;
+	return dummyElement.innerText;
+}
+
+/**
+ * Generates an anchor.
+ *
+ * @param {Object}   block                     The block.
+ * @param {string[]} knownAnchors              The known anchors.
+ * @param {string[]} blocksThatWereNotHeadings The block client IDs that weren't headings in the previous state.
+ * @param {boolean}  fillAllAnchors            Whether or not all empty anchors should be filled.
+ *
+ * @return {string} The anchor.
+ */
+function generateAnchor(
+	block,
+	knownAnchors,
+	blocksThatWereNotHeadings,
+	fillAllAnchors
+) {
+	// Gutenberg doesn't save empty strings.
+	// So when anchor isn't set for a heading that already has content set an empty string.
+	// However, if none of the headings have anchors, assume the page was old, and give all headings an anchor.
+	if (
+		isNil( block.attributes.anchor ) &&
+		! fillAllAnchors &&
+		! isEmpty( block.attributes.content ) &&
+		! blocksThatWereNotHeadings.includes( block.clientId )
+	) {
+		return '';
+	}
+
+	// Get the slug.
+	let slug = cleanForSlug( block.attributes.content );
+	// If slug is empty, then it's likely using non-latin characters.
+	if ( '' === slug ) {
+		slug = trim(
+			deburr( block.attributes.content )
+				.replace( /[\s\./]+/g, '-' )
+				.toLowerCase(),
+			'-'
+		);
+	}
+
+	const baseAnchor = `wp-${ slug }`;
+	let anchor = baseAnchor;
+	let i = 0;
+
+	while ( knownAnchors.includes( anchor ) ) {
+		i += 1;
+		anchor = baseAnchor + '-' + i;
+	}
+
+	return anchor;
+}
+
+/**
+ * Updates the anchor if required.
+ *
+ * @param {Object}   block                     The block.
+ * @param {Object}   knownHeadings             The known headings.
+ * @param {string[]} knownAnchors              The known anchors.
+ * @param {string[]} blocksThatWereNotHeadings The block client IDs that weren't headings in the previous state.
+ * @param {boolean}  fillAllAnchors            Whether or not all empty anchors should be filled.
+ *
+ * @return {string} The anchor.
+ */
+function maybeUpdateAnchor(
+	block,
+	knownHeadings,
+	knownAnchors,
+	blocksThatWereNotHeadings,
+	fillAllAnchors
+) {
+	let anchor = block.attributes.anchor;
+
+	// If the block was previously unknown or has changed content
+	// and the anchor is empty or was auto-generated.
+	if (
+		( ! knownHeadings[ block.clientId ] ||
+			knownHeadings[ block.clientId ].content !==
+				block.attributes.content ) &&
+		( isNil( anchor ) || anchor.startsWith( 'wp-' ) )
+	) {
+		anchor = generateAnchor(
+			block,
+			knownAnchors,
+			blocksThatWereNotHeadings,
+			fillAllAnchors
+		);
+
+		if ( anchor !== block.attributes.anchor ) {
+			dispatch(
+				'core/block-editor'
+			).updateBlockAttributes( block.clientId, { anchor } );
+		}
+	}
+
+	return anchor;
+}
+
+/**
+ * Subscribes to the store to update blocks as they are added or suggestions are updated.
+ *
+ * @return {void}
+ */
+export default function subscribeToStore() {
```

Comment:
Do you think instead of implementing this as a global subscribe, we should just do a react hook on mount of a heading block?

Reply:
Hmmm I wanted it to run not only when a new heading gets added, but also when the content loads...
This way if we open up an older post, all headings will get anchors as they should. It will also make it easier to add a `table-of-contents` block (when that gets refactored and activated) to older posts.
We could do it on mount of a heading block, I'm just not sure that would do what we need to accomplish 🤔

Reply:
> Hmmm I wanted it to run not only when a new heading gets added, but also when the content loads...

so this sounds exactly like a `useEffect(doSomething, [])` in the edit function of the heading block?